### PR TITLE
Add aliasing in deployment

### LINKED ISSUE #1137: 🚧 Dev ➾ Taquito Plugin Improvements (aliases, `call-method`, and `taq transfer`)
### 🔗 Epic|Feature Link: #1167 

https://ecadlabs.productboard.com/feature-board/5219572-taqueria-features/features/14699515

### 🚥 Definition of Done

- [ ] 🎯 Planning and Design
- [ ] 🏎️ Dev Implementation
- [ ] 🏁 Test Plan Executed
- [ ] 🛩️ Documentation Updated
- [ ] 🏟️ Peer / User Reviewed
- [ ] 🏆 Validated and Accepted

### ✅ To Do

_No response_

### 🔭 Scope

🏎️ In Scope: 
 - [ ] Adding new methods to the Taquito plugin
 - [ ] Updating scaffolds
 - [ ] Refactoring Taquito plugin to be more extensible
 - [ ] Implementing contract Aliases

 🙅‍♂️ Out of Scope: 
 - [ ] Work on any other plugins
 - [ ] Removing the State API
 
### 🧰 Work Description

This ticket will implement improvements to the developer workflow by adding adaptive support for the naming convention introduced in the Ligo improvements. Mainly:
- `taq deploy` has an optional argument `--storage / -s` which allows the user to provide a filename to use for the origination
- The `taq deploy` task will look for a storage file that is of format `contractName.DEFAULT_INITIAL_STORAGE.tz` (where `DEFAULT_INITIAL_STORAGE` is "default_storage"). If found, the contents of that file are used for origination

#### New Tasks

The Taquito plugin also needs to have additional tasks added to support a SC dev's workflow. The first two tasks requested are:
- `taq call-method [contractName] [entrypointName]`: Allows the user to call an arbitrary endpoint on a deployed contract
- `taq transfer`: Allows a basic transfer of Tez from one account to another

In order to support a friendly ux for the above methods and to open the door to dApp integration opportunities, `Contract Aliases` will be implemented in this work

#### Contract Aliasing:

Problem Space:

For the purposes of this ticket, a contract alias solves the question: "What is the address of the most recent deployment of a particular contract in the current environment?"

A contract alias is a means for a user to refer to the overall concept of a contract by a single name across multiple deployments. Practically, it is a key:value mapping of an alias name to an object. For now, that object will simply store a value called `address` which will store a KT1 address and should look something like this:

```json
'contractAlias': {
    "address": "KT1..."
}
``` 

This address will be updated each time the user originates a contract to that alias. By default, `taq originate` will assume the alias the user wants is equal to the filename of the contract being originated. The task should accept an optional flag `--alias` which allows the user to explicitly specify an alias to add/update

### ⚖️ Acceptance Criteria

✨ Features: 

#### `taq call-method` task:

- [ ] Taquito plugin exposes a new task to call a method endpoint
- [ ] Entrypoint can be called by annotation
- [ ] Entrypoint can be called with Michelson expression (UNDER EXPLORATION - NOT REQUIRED!)
- [ ] Underlying CLI command needs to support pipes and redirection for use with other commands ie. (- taq compile-expr [expression goes here] | taq call-method [contractName] [entrypointName])
- [ ] The method will accept either a KT1 address, or an alias name for the `[contract]` parameter

#### `taq transfer` task:
- [ ] Should support contract aliases or contract addresses 
- [ ] Can support implicit or originated accounts

#### Contract Aliases:
- [x] Support for aliases added to `config.json`
- [x] The `taq originate` task creates a new alias, or updates one with the same name each time it is called (address is updated to equal the returned address from the underlying call)
- [x] An optional flag `--alias` is added to the `taq originate` task which allows the user to specify a particular alias (creates or updates)

#### 
 
 📚 Documentation: 
- [ ] Revamp the Taquito plugin docs
- [ ] Release notes
 

### 🎢 Test Plan

_No response_

### 🎆 Release Notes

_No response_

### 🐘 Sizing

5

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jchenche**
# 🏗️ PR ➾

Fixes #1137

## 🪂 Pre-Merge Checklist

- [x] 🐬 I have commented my code, particularly in hard-to-understand areas
- [ ] 🦀 Automated tests have been written and added to this PR
- [ ] 🤿 Corresponding changes have been made to all documentation
- [ ] 🏖️ New and existing unit tests pass locally and in CI
- [ ] 🔱 The test plan has been implemented and verified by an SDET
- [x] 🏄‍♂️ Myself and others have built this branch and done manual testing on the change
- [x] ⛵ A summary describing the change has been written for release notes

----------------------------------------------------------------------------------------------------------------------------

## 🪁 Description

This extends the config environment with an `aliases` field that stores aliases to deployed contract's addresses. This feature enables hot reloading because every time we deploy, it will update the address mapped by the alias.

## 🎢 Test Plan

Please describe the testing strategy for this change

----------------------------------------------------------------------------------------------------------------------------

### 🛸 Type of Change

- [ ] 🐟 Minor change (non-breaking change of very limited scope)
- [ ] 🦑 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🌊 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🐳 Major refactor (breaking changes and significant impact to the codebase)

----------------------------------------------------------------------------------------------------------------------------

### CODE REVIEW

**Comment by hu3man** [comment]
@alexzbusko @jchenche , what's the test plan (if any) for this?

**Comment by hu3man** [comment]
I think this error message is outdated since we aren't looking for the contract location in `config.json` anymore

```
❯ taq originate test.tz
Michelson artifact test.tz has no initial storage specified for the target environment.
Storage is expected to be specified in .taq/config.json at JSON path: environment.development.storage["test.tz"]
The value of the above JSON key should be the name of the file (absolute path or relative path with respect to the root of the Taqueria project) that contains the actual value of the storage, as a Michelson expression.

No contracts originated.
```

**Comment by jchenche** [comment]
> @alexzbusko @jchenche , what's the test plan (if any) for this?

Haven't come up with one yet but I guess it would be something like:
Test where aliases field doesn't exist.
Test where the specific alias field doesn't exist.
Test where the address field doesn't exist.
Test with option --alias
Test that the value of the address field is updated after each deployment.

All of which I manually tested.

**Comment by hu3man** [comment]
Great, thanks @jchenche. I'll leave it to you and @alexzbusko to determine the appropriate amount of test coverage for the short term (keeping in mind upcoming Environment refactor)

File: taqueria-plugin-taquito/originate.ts
Comment (hu3man) [inline taqueria-plugin-taquito/originate.ts:79]: I like the additional actionable context given to the user here. 

Suggestions for clarity:
```
❌ No initial storage file was found for this contract

Storage must be specified in a file as a Michelson expression and will automatically be linked to this contract if specified in ${storageFilename} in the artifacts dir

You can also pass the filename to the deploy task using the --storage STORAGE_FILE_NAME option
```

**Review by hu3man — APPROVED** [review]
Excellent work and appreciate the fast time to market!

One comment regarding error output but has my approval other than that

File: taqueria-plugin-taquito/originate.ts
Comment (jchenche) [reply taqueria-plugin-taquito/originate.ts:79]: Thanks! Much clearer! I will copy this verbatim 😅

**Review by sinapsist — APPROVED** [review]
LGTM

**Pull request merged**
